controllers/dscinitialization/utils.go

```diff
@@ -18,132 +21,178 @@ import (
+	err = r.reconcileDefaultNetworkPolicy(ctx, nsName, dscInit, platform, log)
 	if err != nil {
-		log.Error(err, "error reconciling network policy ", "name", name)
+		log.Error(err, "error reconciling network policy ", "name", nsName)
 		return err
 	}
 
 	// Create odh-common-config Configmap for the Namespace
-	err = r.createOdhCommonConfigMap(ctx, name, dscInit)
+	err = r.createOdhCommonConfigMap(ctx, nsName, dscInit, log)
 	if err != nil {
 		log.Error(err, "error creating configmap", "name", "odh-common-config")
 		return err
 	}
 
 	// Create default Rolebinding for the namespace
-	err = r.createDefaultRoleBinding(ctx, name, dscInit)
+	err = r.createDefaultRoleBinding(ctx, nsName, dscInit, log)
 	if err != nil {
-		log.Error(err, "error creating rolebinding", "name", name)
+		log.Error(err, "error creating rolebinding", "name", nsName)
 		return err
 	}
 	return nil
 }
 
-func (r *DSCInitializationReconciler) createDefaultRoleBinding(ctx context.Context, name string, dscInit *dsciv1.DSCInitialization) error {
-	log := logf.FromContext(ctx)
+func (r *DSCInitializationReconciler) appNamespaceHandler(ctx context.Context, nsName string, log logr.Logger) error {
+	// Check if application namespace has label "opendatahub.io/application-namespace:true"
+	// if no such namespace exist, we create it with generated-namespace and security label
+	// if namespace exist but no label, we exit
+	// if namespace exist and has label, we enforce security label onto it.
+	appNamespace := &corev1.Namespace{}
+	desiredAppNS := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nsName,
+		},
+	}
+	if err := r.Get(ctx, client.ObjectKeyFromObject(desiredAppNS), appNamespace); err != nil {
+		if !k8serr.IsNotFound(err) {
+			return err
+		}
+		return r.createAppNamespace(ctx, nsName)
+	}
+
+	l := appNamespace.GetLabels()
+	if l[labels.CustomizedAppNamespace] == "true" {
+		resources.SetLabel(appNamespace, labels.SecurityEnforce, "baseline")
+		if err := r.Update(ctx, appNamespace); err != nil {
+			log.Error(err, "Failed to force security label on namespace", "name", nsName)
+		}
+	}
+	return errors.New("application namespace missing required label or label value is incorrect. Please recreate DSCI to set the label, or leave it to default value")
+}
+
+func (r *DSCInitializationReconciler) createAppNamespace(ctx context.Context, nsName string) error {
+	desiredDefaultNS := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: nsName},
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, r.Client, desiredDefaultNS, func() error {
+		resources.SetLabel(desiredDefaultNS, labels.ODH.OwnedNamespace, labels.True) // this indicate when uninstall, namespace will be deleted
+		resources.SetLabel(desiredDefaultNS, labels.SecurityEnforce, "baseline")
+		return nil
+	})
+	return err
+}
+
+func (r *DSCInitializationReconciler) patchAppNS(ctx context.Context, nsName string, log logr.Logger) error {
```

Comment:
can this part of the create app namespace ? we could avoid a number of calls by properly setting up the namespace at creation/update time